### PR TITLE
arm: implement API to validate user buffer

### LINKED ISSUE #3832: ARM: implement API to validate user buffer
**_Reported by Andrew Boie:_**

Implement a function:

```

#define BUF_READABLE
#define BUF_WRITABLE
#define BUF_USER

int _arm_mpu_buffer_validate(void *addr, size_t size, int flags);
```

Such that for the entire memory region defined by addr/size, that the current MPU configuration will allow reads or writes to that memory. If the BUF_USER flag is set, the permissions need to extend to non-supervisor mode, although you may assume that this function will only be run in supervisor mode.

This will be used in system calls where the end user passes in buffers, we need to make sure that any buffers are in memory that the calling thread actually has access to.

(Imported from Jira ZEP-2399)

### PR DESCRIPTION
**Pull request by chunlinhan**
Implement API to validate user buffer. This API will iterate
all MPU regions to check if the given buffer is user accessible
or not. 

Fixes #3832.

Signed-off-by: Chunlin Han <chunlin.han@linaro.org>

### CODE REVIEW

File: arch/arm/core/cortex_m/mpu/nxp_mpu.c [hunk 1]
```diff
@@ -112,6 +112,46 @@ static inline u32_t _get_region_index_by_type(u32_t type)
 	}
 }
 
+/**
+ * This internal function check if region is enabled or not
+ */
+static inline int _is_enabled_region(u32_t r_index)
+{
+	return SYSMPU->WORD[r_index][3] & SYSMPU_WORD_VLD_MASK;;
```
Comment (galak) [inline arch/arm/core/cortex_m/mpu/nxp_mpu.c]: nit double ;;

**Review by galak — CHANGES_REQUESTED** [review]
minor comment, but looks good to me.  Would be good to add more to the commit message about the change.

**Review by agross-oss — COMMENTED** [review]
Just update the commit message to be more descriptive.  The rest looks good to me.

**Review by tyrantlizard77 — APPROVED** [review]

File: arch/arm/core/cortex_m/mpu/nxp_mpu.c (hunk 1 above)
Comment (chunlinhan) [reply arch/arm/core/cortex_m/mpu/nxp_mpu.c]: fixed.

**Comment by chunlinhan** [comment]
Update commit message.

**Comment by chunlinhan** [comment]
fix typo.

**Review by nashif — APPROVED** [review]

**Pull request merged**